src/Features/Core/Portable/Completion/Providers/Snippets/AbstractSnippetCompletionProvider.cs

```diff
@@ -0,0 +1,119 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Snippets;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.Completion.Providers.Snippets
+{
+    internal abstract class AbstractSnippetCompletionProvider : CompletionProvider
+    {
+        public AbstractSnippetCompletionProvider()
+        {
+
+        }
+
+        internal abstract string Language { get; }
+
+        public override async Task<CompletionChange> GetChangeAsync(Document document, CompletionItem item, char? commitKey = null, CancellationToken cancellationToken = default)
+        {
+            var (strippedDocument, position) = await GetDocumentWithoutInvocationAsync(document, SnippetCompletionItem.GetInvocationPosition(item), cancellationToken).ConfigureAwait(false);
+            var service = strippedDocument.GetRequiredLanguageService<ISnippetService>();
+            var snippetProvider = service.GetSnippetProvider(new SnippetData(item.DisplayText));
+            var snippet = await snippetProvider.GetSnippetAsync(strippedDocument, position, cancellationToken).ConfigureAwait(false);
+            var strippedText = await strippedDocument.GetTextAsync(cancellationToken).ConfigureAwait(false);
+            var allChangesText = strippedText.WithChanges(snippet.TextChanges);
+            var allChangesDocument = document.WithText(allChangesText);
+            var allTextChanges = await allChangesDocument.GetTextChangesAsync(document, cancellationToken).ConfigureAwait(false);
+            var change = Utilities.Collapse(allChangesText, allTextChanges.AsImmutable());
+            return CompletionChange.Create(change, newPosition: snippet.CursorPosition, includesCommitCharacter: true);
```

Comment:
i think you want to pass along the completion change that has the full change and the small changes too.

Reply:
nit: comment what's going on here.  it's useful to do it with a running example to help motivate why we're doing each step (or chunk of steps).

Reply:
above comment still applies.